File: firebase/functions/lib/events/live_meetings/breakouts/check_hostless_go_to_breakouts.dart
import 'dart:async';

import 'package:firebase_functions_interop/firebase_functions_interop.dart';
import 'initiate_breakouts.dart';
import '../../../on_call_function.dart';
import 'check_hostless_go_to_breakouts_server.dart';
import '../../../utils/firestore_utils.dart';
import 'package:data_models/cloud_functions/requests.dart';
import 'package:data_models/firestore/event.dart';
import 'package:data_models/firestore/live_meeting.dart';

class CheckHostlessGoToBreakouts
    extends OnCallMethod<CheckHostlessGoToBreakoutsRequest> {
  CheckHostlessGoToBreakouts()
      : super(
          CheckHostlessGoToBreakoutsRequest.functionName,
          (jsonMap) => CheckHostlessGoToBreakoutsRequest.fromJson(jsonMap),
          runWithOptions: RuntimeOptions(
            timeoutSeconds: 240,
            memory: '4GB',
            minInstances: 0,
          ),
        );

  @override
  Future<void> action(
    CheckHostlessGoToBreakoutsRequest request,
    CallableContext context,
  ) async {
    await checkHostlessGoToBreakouts(request, context.authUid!);
  }

  Future<void> checkHostlessGoToBreakouts(
    CheckHostlessGoToBreakoutsRequest request,
    String userId,
  ) async {
    final event = await firestoreUtils.getFirestoreObject(
      path: request.eventPath,
      constructor: (map) => Event.fromJson(map),
    );
    print('Retrieved event: $event');

    if (event.eventType != EventType.hostless) {
      print('Event is not hostless, returning.');
      return;
    }

    if (event.status != EventStatus.active) {
      print('Event is cancelled, returning.');
      return;
    }

    final nowWithBuffer = DateTime.now().add(const Duration(milliseconds: 100));

    final timeUntilWaitingRoomFinished =
        event.timeUntilWaitingRoomFinished(nowWithBuffer);

    print(
        'comparing now ($nowWithBuffer) to scheduled start time (${event.scheduledTime}). '
        'Waiting room finished in $timeUntilWaitingRoomFinished');
    if (!timeUntilWaitingRoomFinished.isNegative) {
      print('It is still before scheduled start time so returning.');
      return;
    }

    final liveMeetingPath = '${event.fullPath}/live-meetings/${event.id}';
    final liveMeeting = await firestoreUtils.getFirestoreObject(
      path: liveMeetingPath,
      constructor: (map) => LiveMeeting.fromJson(map),
    );

    if ([BreakoutRoomStatus.active]
        .contains(liveMeeting.currentBreakoutSession?.breakoutRoomStatus)) {
      print('Breakouts have already been assigned so returning.');
      return;
    }
    const defaultTargetParticipants = 8;

    print('initializing breakouts');
    await InitiateBreakouts().initiateBreakouts(
      event: event,
      request: InitiateBreakoutsRequest(
        eventPath: event.fullPath,
        // Use the event ID so that any duplicate checks will use the same breakout session ID.
        breakoutSessionId: event.id,
        assignmentMethod: event.breakoutRoomDefinition?.assignmentMethod ??
            BreakoutAssignmentMethod.targetPerRoom,
        targetParticipantsPerRoom:
            event.breakoutRoomDefinition?.targetParticipants ??
                defaultTargetParticipants,
        includeWaitingRoom: true,
      ),
      creatorId: userId,
    );
    print('Finished initiating breakouts');
  }

  Future<void> enqueueScheduledCheck(Event event) async {
    final timeToGoToBreakouts =
        DateTime.now().add(event.timeUntilWaitingRoomFinished(DateTime.now()));

    await CheckHostlessGoToBreakoutsServer().schedule(
      CheckHostlessGoToBreakoutsRequest(eventPath: event.fullPath),
      timeToGoToBreakouts,
    );
  }
}

File: firebase/functions/src/events/live_meetings/breakouts/check_hostless_go_to_breakouts.ts
import * as functions from 'firebase-functions';
import { OnCallMethod } from '../../../on_call_function';
import { firestore, firestoreUtils } from '../../../utils/infra/firestore_utils';
import { Event, EventType, EventStatus, BreakoutRoomStatus, BreakoutAssignmentMethod, LiveMeeting, timeUntilWaitingRoomFinished, getEventFullPath } from '../../../types';
import { InitiateBreakouts } from './initiate_breakouts';
import { CheckHostlessGoToBreakoutsServer } from './check_hostless_go_to_breakouts_server';

interface CheckHostlessGoToBreakoutsRequest {
  eventPath: string;
}

export class CheckHostlessGoToBreakouts extends OnCallMethod<CheckHostlessGoToBreakoutsRequest> {
  constructor() {
    super(
      'CheckHostlessGoToBreakouts',
      (jsonMap) => jsonMap as CheckHostlessGoToBreakoutsRequest,
      { runWithOptions: { timeoutSeconds: 240, memory: '4GB', minInstances: 0 } }
    );
  }

  async action(request: CheckHostlessGoToBreakoutsRequest, context: functions.https.CallableContext): Promise<void> {
    await this.checkHostlessGoToBreakouts(request, context.auth!.uid);
  }

  async checkHostlessGoToBreakouts(request: CheckHostlessGoToBreakoutsRequest, userId: string): Promise<void> {
    const event: Event = await firestoreUtils.getFirestoreObject({
      path: request.eventPath,
      constructor: (map) => map as unknown as Event,
    });
    console.log('Retrieved event:', (event as any).id);

    if ((event as any).eventType !== EventType.hostless) {
      console.log('Event is not hostless, returning.');
      return;
    }

    if ((event as any).status === EventStatus.canceled) {
      console.log('Event is cancelled, returning.');
      return;
    }

    const nowWithBuffer = new Date(Date.now() + 100);
    const timeUntilWaitingRoom = timeUntilWaitingRoomFinished(event, nowWithBuffer);

    console.log(`Waiting room finished in ${timeUntilWaitingRoom}ms`);
    if (timeUntilWaitingRoom >= 0) {
      console.log('It is still before scheduled start time so returning.');
      return;
    }

    const liveMeetingPath = `${getEventFullPath(event)}/live-meetings/${(event as any).id}`;
    const liveMeeting: LiveMeeting = await firestoreUtils.getFirestoreObject({
      path: liveMeetingPath,
      constructor: (map) => map as unknown as LiveMeeting,
    });

    if ((liveMeeting as any).currentBreakoutSession?.breakoutRoomStatus === BreakoutRoomStatus.active) {
      console.log('Breakouts have already been assigned so returning.');
      return;
    }

    const defaultTargetParticipants = 8;
    console.log('initializing breakouts');
    await new InitiateBreakouts().initiateBreakouts({
      event,
      request: {
        eventPath: getEventFullPath(event),
        breakoutSessionId: (event as any).id,
        assignmentMethod: (event as any).breakoutRoomDefinition?.assignmentMethod ?? BreakoutAssignmentMethod.targetPerRoom,
        targetParticipantsPerRoom: (event as any).breakoutRoomDefinition?.targetParticipants ?? defaultTargetParticipants,
        includeWaitingRoom: true,
      },
      creatorId: userId,
    });
    console.log('Finished initiating breakouts');
  }

  async enqueueScheduledCheck(event: Event): Promise<void> {
    const timeToGoToBreakoutsMs = timeUntilWaitingRoomFinished(event, new Date());
    const timeToGoToBreakouts = new Date(Date.now() + timeToGoToBreakoutsMs);

    await new CheckHostlessGoToBreakoutsServer().schedule(
      { eventPath: getEventFullPath(event) },
      timeToGoToBreakouts
    );
  }
}
